#COMISION 25262 - DATA ANALYTICS

>  ## ETAPA:  *"Pre Entrega Proyecto"*.
   **Alumno**: Javier Marcelo Espiñeira (DNI 23.146.908).



## Guía para el lector / evaluador. (Sobre enunciado original propuesto)
*   **Sección 1:** Recopilación y Preparación de Datos (Clases 1 a 4)
    *   ***Objetivo:** Demostrar habilidades en Python, familiaridad con el entorno de trabajo y conocimientos básicos sobre manipulación de datos.*
    *   **Actividades realizadas:**

        1. Crear un documento en Google Colaboratory y cargar los sets de datos como DataFrames.<br>
        👉 Ir a *"C) Carga de los datos de entrada como DataFrames."* usando el índice lateral.

        2. Realizar un script básico que calcule las ventas mensuales utilizando variables y operadores.<br>
        👉 Punto ELIMINADO del temario.

        3. Estructuras de Datos: Desarrollar un programa que almacene los datos de ventas (producto, precio, cantidad). Decidir si conviene utilizar diccionarios o listas.<br>
        👉 Punto ELIMINADO del temario.

        4. Introducción a Pandas: realizar un análisis exploratorio inicial de los DataFrames.<br>
        👉 Ir a *"D) Exploración inicial de los datos."* usando el índice lateral.

        5. Calidad de Datos: Identificar valores nulos y duplicados en los conjuntos de datos. Documentar el estado inicial de los datos.<br>
        👉 Ir a *"F) Detección y tratamiento de valores nulos"., G) Detección y tratamiento de duplicados. y H)a) Copia / reserva de los DataFrames originales.* usando el índice lateral.
         <br>

*   **Sección 2:** Pre procesamiento y Limpieza de Datos (Clases 5 a 8)

    *   ***Objetivo:** Demostrar conocimiento de las técnicas de limpieza y transformación de datos.*

    *   **Actividades realizadas:**

        1. Limpieza de Datos: Limpiar el conjunto de datos eliminando duplicados y caracteres no deseados. Documentar el proceso y los resultados.<br>
        👉 Ir a *"H) Limpieza y transformación de los DataFrames"* usando el índice lateral.

        2. Transformación de Datos: Aplicar filtros y transformaciones para crear una tabla de ventas que muestre solo los productos con alto rendimiento.<br>
        👉 Ir a *"I) Enriquesimiento de datos, Anexo Especial a)"* usando el índice lateral.

        3. Agregación: Resumir las ventas por categoría de producto y analizar los ingresos generados.<br>
        👉 Ir a *"I) Enriquesimiento de datos, Anexo Especial b)"* usando el índice lateral.

        4. Integración de Datos: Combinar los sets de datos de ventas y marketing para obtener una visión más amplia de las tendencias.<br>
        👉 Punto Desafectado de esta entrega.
        <br>

*   **Anexo:**

    *   ***Objetivo:** Descripción y links a todos los archivos adicionales.*

    *    Descripciones de archivos y links.<br>
         👉 Tarea Descontinuada.
  <br>





# A) Preparación del entorno y carga de librerías utilizadas.

In [39]:
# Importar la librería Pandas y NumPy
import pandas as pd
import numpy as np
import re

# B) Funciones utilizadas.
>Se mantienen por separado las funciones utilizadas tanto para su futura reutilización, testeo y mantenimiento como para diferenciar la logica de funcionamiento interno de cada función con el proposito que esta cumple en el proceso que se está realizando con los datos.
>Si bien no se detectaron errores para lo realizado en este caso en particular, mi idea es a futuro agregar testeo para mejorar la confianza en las mismas y ganar confiabilidad en explorar la mayoría de los caminos y ocurrencias posibles. (de hecho se observaron problemas principalmente en los mensajes personalizados de error al parecer entiendo por manejo de errores la plataforma Colab -en Python puro funcionan y aquí, no por ejemplo, pero se dejó para más adelante estos temas que exceden el trabajo solicitado)  

In [40]:
# Funciones de Uso General

def generar_url_csv_gdrive(ID_planilla):
  """
  Genera una URL de CSV en Google Drive a partir de un ID de hoja.(Solo compartido o accesible para todos con enlace)

  Args:
    sheet_id: El ID del documento de Google Drive.

  Returns:
    La URL de exportación CSV para ser utilizada en pandas ".read_csv"
  """
  url = f'https://drive.google.com/uc?export=download&id={ID_planilla}'
  return url


def realizar_exploracion_inicial_ejemplo_visual(DF, nombre="", cantidad_lineas=5):
  """
  Realiza una exploración inicial visual de un DataFrame, mostrando las primeras, últimas y filas de ejemplo.

  Args:
    DF: El DataFrame a explorar.
    nombre: Un nombre para el DataFrame (opcional).
    cantidad_lineas: La cantidad de líneas a mostrar en head, tail y sample (opcional).
  """
  if DF is None:
    print(f"Error: El DataFrame '{nombre}' no ha sido definido o es None.")
    return

  print(f"\n\n '{nombre}' Primeras filas:\n")
  print(DF.head(cantidad_lineas))
  print(f"\n\n '{nombre}' Filas de ejemplo -al azar-:\n")
  print(DF.sample(cantidad_lineas))
  print(f"\n\n '{nombre}' Últimas filas:\n")
  print(DF.tail(cantidad_lineas))


def realizar_exploracion_inicial_profunda(DF, nombre=""):
  """
  Realiza una exploración inicial profunda de un DataFrame, Analizando su forma,
  tamaño caracteristicas de sus datos e información estadística de los mismos.

  Args:
    DF: El DataFrame a explorar.
    nombre: Un nombre para el DataFrame (opcional).
  """
  if DF is None:
    print(f"Error: El DataFrame '{nombre}' no ha sido definido o es None.")
    return

  print(f"\n\n '{nombre}' Analisis de Tamaño: (Filas, Columnas)\n")
  dimensiones = DF.shape
  print(dimensiones)
  print(f"\n\n '{nombre}' Tipos de Datos, Nulos, etc.:\n")
  print(DF.info())
  print(f"\n\n '{nombre}' Informacion Estadistica del set:\n")
  print(DF.describe(include= "all"))


def contar_nulos_por_columna(DF, nombre=""):
  """
  Cuenta y muestra la cantidad de valores nulos por columna en un DataFrame.

  Args:
    DF: El DataFrame a analizar.
    nombre: Un nombre para el DataFrame (opcional)
  """
  if DF is None:
    print("Error: El DataFrame no ha sido definido o es None.")
    return

  print(f"\nConteo de valores nulos por columna del dataframe '{nombre}':\n")
  for columna in DF.columns:
    nulos = DF[columna].isnull().sum()
    print(f"Columna '{columna}': {nulos} valores nulos")


def mostrar_nulos_de_una_columna(DF, columna, nombre=""):
  """
  Muestra las filas de un DataFrame con valores nulos en una columna específica.

  Args:
    DF: El DataFrame a analizar.
    columna: El nombre de la columna a verificar en busca de nulos.
    nombre: Un nombre para el DataFrame (opcional)
  """
  if DF is None:
    print("Error: El DataFrame no ha sido definido o es None.")
    return

  if columna not in DF.columns:
    print(f"Error: La columna '{columna}' no existe en el DataFrame '{nombre}'.")
    return

  df_filas_con_nulos = DF[DF[columna].isnull()]
  print(f"\nFilas con valores nulos en la columna '{columna}' del dataframe '{nombre}':\n")
  print(df_filas_con_nulos.head(len(df_filas_con_nulos)))


def detectar_duplicados(DF, nombre="", columnas_pk=None):
  """
  Detecta y muestra las filas duplicadas en un DataFrame.

  Args:
    DF: El DataFrame a analizar.
    nombre: Un nombre para el DataFrame (opcional)
    columnas_pk: Lista de columnas a considerar como clave primaria (opcional)
  """
  if DF is None:
    print("Error: El DataFrame no ha sido definido o es None.")
    return

  if columnas_pk:
    print(f"\nConteo de valores duplicados en el dataframe '{nombre}' por {columnas_pk}:\n")
    duplicados = DF.duplicated(subset=columnas_pk).sum()
    print(f"Cantidad de filas duplicadas: {duplicados}")

    if duplicados > 0:
      print(f"\nFilas duplicadas en el dataframe '{nombre}' por {columnas_pk}:\n")
      duplicated_rows = DF[DF.duplicated(subset=columnas_pk, keep=False)]
      display(duplicated_rows.groupby(columnas_pk).size().reset_index(name='cantidad_duplicados'))

  else:
    print(f"\nConteo de valores duplicados en el dataframe '{nombre}':\n")
    duplicados = DF.duplicated().sum()
    print(f"Cantidad de filas duplicadas: {duplicados}")

    if duplicados > 0:
      print(f"\nFilas duplicadas en el dataframe '{nombre}':\n")
      display(DF[DF.duplicated()])


def eliminar_duplicados(DF, nombre="", columnas_pk=None, tipo_de_eliminacion='first'):
  """
  Elimina las filas duplicadas de un DataFrame.

  Args:
    DF: El DataFrame del que se eliminarán los duplicados.
    nombre: Un nombre para el DataFrame (opcional)
    columnas_pk: Lista de columnas a considerar como clave primaria (opcional).
                 Si es None, se consideran todas las columnas.
    tipo_de_eliminacion: Determina qué duplicados eliminar. 'first' elimina todos excepto la primera aparición,
                         'last' elimina todos excepto la última, False elimina todos los duplicados. (opcional, por defecto 'first')

  Returns:
    El DataFrame sin duplicados.
  """
  if DF is None:
    print("Error: El DataFrame no ha sido definido o es None.")
    return None

  filas_al_inicio = DF.shape[0]

  if columnas_pk:
    df_modificado = DF.drop_duplicates(subset=columnas_pk, keep=tipo_de_eliminacion)
    print(f"\nDuplicados eliminados del dataframe '{nombre}' por {columnas_pk}:")
  else:
    df_modificado = DF.drop_duplicates(keep=tipo_de_eliminacion)
    print(f"\nDuplicados eliminados del dataframe '{nombre}':")

  filas_eliminadas = filas_al_inicio - df_modificado.shape[0]
  print(f"Cantidad de filas eliminadas: {filas_eliminadas}")

  return df_modificado


def intervenir_nulos_de_una_columna(DF, columna, tipo_operacion, valor_personalizado=None, nombre=""):
  """
  Interviene en los valores nulos de una columna específica de un DataFrame
  basado en el tipo de operación especificada.

  Args:
    DF: El DataFrame a modificar.
    columna: El nombre de la columna donde se intervendrán los nulos.
    tipo_operacion: El tipo de operación a realizar. Puede ser 'eliminar',
                    'media', 'moda', 'mediana', o 'personalizado'.
    valor_personalizado: El valor a utilizar si tipo_operacion es 'personalizado'.
                       (Opcional, por defecto "")
    nombre: Un nombre para el DataFrame (opcional)

  Returns:
    El DataFrame modificado.
  """
  if DF is None:
    print("Error: El DataFrame no ha sido definido o es None.")
    return None

  if columna not in DF.columns:
    print(f"Error: La columna '{columna}' no existe en el DataFrame.")
    return DF

  df_modificado = DF.copy()
  nulos_iniciales = df_modificado[columna].isnull().sum()

  if nulos_iniciales == 0:
    print(f"No hay valores nulos en la columna '{columna}' del dataframe '{nombre}'.")
    return df_modificado

  print(f"\nInterviniendo valores nulos en la columna '{columna}' del dataframe '{nombre}' con operación '{tipo_operacion}':")

  if tipo_operacion == 'eliminar':
    df_modificado = df_modificado.dropna(subset=[columna])
    print(f"Se eliminaron las filas con valores nulos en la columna '{columna}'.")
  elif tipo_operacion == 'media':
    try:
      media = df_modificado[columna].mean()
      df_modificado[columna] = df_modificado[columna].fillna(media)
      print(f"Los valores nulos en la columna '{columna}' fueron reemplazados por la media ({media}).")
    except TypeError:
      print(f"Error: La columna '{columna}' no es numérica, no se puede calcular la media.")
  elif tipo_operacion == 'moda':
    try:
      moda = df_modificado[columna].mode()[0]
      df_modificado[columna] = df_modificado[columna].fillna(moda)
      print(f"Los valores nulos en la columna '{columna}' fueron reemplazados por la moda ({moda}).")
    except Exception as e:
        print(f"Error al calcular la moda para la columna '{columna}': {e}")
  elif tipo_operacion == 'mediana':
    try:
      mediana = df_modificado[columna].median()
      df_modificado[columna] = df_modificado[columna].fillna(mediana)
      print(f"Los valores nulos en la columna '{columna}' fueron reemplazados por la mediana ({mediana}).")
    except TypeError:
      print(f"Error: La columna '{columna}' no es numérica, no se puede calcular la mediana.")
  elif tipo_operacion == 'personalizado':
    if valor_personalizado is not None:
      df_modificado[columna] = df_modificado[columna].fillna(valor_personalizado)
      print(f"Los valores nulos en la columna '{columna}' fueron reemplazados por el valor personalizado ({valor_personalizado}).")
    else:
      print("Error: Se especificó 'personalizado' pero no se proporcionó 'valor_personalizado'.")
  else:
    print(f"Error: Tipo de operación '{tipo_operacion}' no reconocido.")

  nulos_finales = df_modificado[columna].isnull().sum()
  filas_afectadas = nulos_iniciales - nulos_finales
  print(f"Cantidad de valores nulos antes: {nulos_iniciales}, después: {nulos_finales}. Filas afectadas: {filas_afectadas}")


  return df_modificado


def normalizar_texto(DF, columnas, nombre="", eliminar_espacios_extra=True, eliminar_caracteres_especiales=True, tipo_conversion=None, mapeo_caracteres_especiales=None):
  """
  Normaliza el texto en columnas específicas de un DataFrame.

  Args:
    DF: El DataFrame a modificar.
    columnas: Una lista de nombres de columnas a normalizar.
    nombre: Un nombre para el DataFrame (opcional).
    eliminar_espacios_extra: Booleano, si True elimina espacios en blanco adicionales (inicio, fin, y multiples intermedios).
    eliminar_caracteres_especiales: Booleano, si True elimina caracteres especiales definidos.
    tipo_conversion: Cadena, puede ser 'title', 'lower', 'upper', o None para no cambiar el caso.
    mapeo_caracteres_especiales: Diccionario opcional para mapear caracteres especiales a reemplazos.
                      Por defecto, mapea '\u200b\t\r\n' a '' y 'ñ' a 'nh', vocales acentuadas a sin acentuar, y apostrofes a ''.

  Returns:
    El DataFrame modificado con el texto normalizado en las columnas especificadas.
  """
  if DF is None:
    print("Error: El DataFrame no ha sido definido o es None.")
    return None

  df_modificado = DF.copy()

  if mapeo_caracteres_especiales is None:
      mapeo_caracteres_especiales = {
          'ñ': 'nh', 'Ñ': 'Nh',
          'á': 'a', 'é': 'e', 'í': 'i', 'ó': 'o', 'ú': 'u',
          'Á': 'A', 'É': 'E', 'Í': 'I', 'Ó': 'O', 'Ú': 'U',
          "'": "" # para quitar apostrofes simples intermedios
          }


  for columna in columnas:
    if columna not in df_modificado.columns:
      print(f"Advertencia: La columna '{columna}' no existe en el DataFrame '{nombre}'. Se omite.")
      continue

    # Convertir a string
    df_modificado[columna] = df_modificado[columna].astype(str)

    # Eliminar espacios en blanco adicionales, tabulaciones, saltos de página, etc.
    if eliminar_espacios_extra:
      df_modificado[columna] = df_modificado[columna].str.strip().str.replace(r'\s+', ' ', regex=True).replace(r"[\u200b\t\r\n]", "", regex=True)


    # Eliminar caracteres especiales
    if eliminar_caracteres_especiales:
        for char, replacement in mapeo_caracteres_especiales.items():
            df_modificado[columna] = df_modificado[columna].str.replace(char, replacement, regex=False)


    # Convertir caso
    if tipo_conversion == 'title':
      df_modificado[columna] = df_modificado[columna].str.title()
    elif tipo_conversion == 'lower':
      df_modificado[columna] = df_modificado[columna].str.lower()
    elif tipo_conversion == 'upper':
      df_modificado[columna] = df_modificado[columna].str.upper()
    elif tipo_conversion is not None:
        print(f"Advertencia: Valor no válido para 'convertir_caso' en la columna '{columna}'. Debe ser 'title', 'lower', 'upper', o None.")
  print(f"\n Normalización de texto aplicada a las columnas {columnas} del dataframe '{nombre}'.")
  return df_modificado


def convertir_formato_fecha(DF, columnas, nombre="", formato_original="%d/%m/%Y"):
    """
    Convierte las columnas de fecha de un DataFrame a tipo datetime64.

    Args:
      DF: El DataFrame a modificar.
      columnas: Una lista de nombres de columnas con tipo de datos de fecha a intervenir.
      nombre: Un nombre para el DataFrame (opcional).
      formato_original: El formato actual de las fechas en las columnas (por defecto "%d/%m/%Y").

    Returns:
      El DataFrame modificado con las columnas de fecha convertidas.
    """
    if DF is None:
        print("Error: El DataFrame no ha sido definido o es None.")
        return None

    df_modificado = DF.copy()

    print(f"\nConvirtiendo formato de fecha y tipo de datos a datetime64 en el dataframe '{nombre}':")

    for columna in columnas:
        if columna not in df_modificado.columns:
            print(f"Advertencia: La columna '{columna}' no existe en el DataFrame '{nombre}'. Se omite.")
            continue

        try:
            # Convertir a datetime con el formato original
            df_modificado[columna] = pd.to_datetime(df_modificado[columna], format=formato_original, errors='coerce')

            print(f"Columna '{columna}' convertida a datetime64.")

        except ValueError as e:
            print(f"Error al convertir la columna '{columna}': {e}")
            print(f"Asegúrese de que el formato original '{formato_original}' coincide con los datos.")
        except Exception as e:
            print(f"Ocurrió un error inesperado al procesar la columna '{columna}': {e}")

    print(f"\n Conversión de formato de fecha y tipo de datos completada para las columnas {columnas} del dataframe '{nombre}'.")

    return df_modificado


def convertir_formato_numerico(DF, columnas, nombre="", caracteres_a_eliminar=None, tipo_numerico='float64'):
    """
    Convierte columnas de un DataFrame con datos numéricos en formato de texto a tipo numérico.

    Args:
      DF: El DataFrame a modificar.
      columnas: Una lista de nombres de columnas a convertir.
      nombre: Un nombre para el DataFrame (opcional).
      caracteres_a_eliminar: Lista de caracteres a eliminar antes de la conversión (por defecto None).
                             Si es None, elimina '$', ',', y espacios en blanco.
      tipo_numerico: El tipo numérico deseado ('int64' o 'float64', por defecto 'float64').

    Returns:
      El DataFrame modificado con las columnas convertidas a tipo numérico.
    """
    if DF is None:
        print("Error: El DataFrame no ha sido definido o es None.")
        return None

    df_modificado = DF.copy()

    if caracteres_a_eliminar is None:
      caracteres_a_eliminar = ['$', ',', ' '] # Caracteres comunes a eliminar

    print(f"\nConvirtiendo columnas a formato numérico en el dataframe '{nombre}':")

    for columna in columnas:
        if columna not in df_modificado.columns:
            print(f"Advertencia: La columna '{columna}' no existe en el DataFrame '{nombre}'. Se omite.")
            continue

        try:
            # Convertir a string para asegurar que se puedan aplicar los métodos de cadena
            df_modificado[columna] = df_modificado[columna].astype(str)

            # Eliminar caracteres no deseados
            for char in caracteres_a_eliminar:
                df_modificado[columna] = df_modificado[columna].str.replace(char, '', regex=False)

            # Manejar el separador decimal si no es un punto (ej. coma)
            # Este ejemplo asume que después de eliminar miles y símbolos,
            # el único separador decimal potencial que queda es una coma si no es punto.
            # Si tus datos tienen otros separadores decimales, ajusta aquí.
            df_modificado[columna] = df_modificado[columna].str.replace(',', '.', regex=False)


            # Convertir a tipo numérico
            if tipo_numerico == 'float64':
                df_modificado[columna] = pd.to_numeric(df_modificado[columna], errors='coerce')
            elif tipo_numerico == 'int64':
                # Para convertir a int, primero convertimos a float para manejar decimales
                # y luego a int. Los valores que no se puedan convertir a float o que
                # tengan decimales después de la limpieza y deban ser int resultarán en NaN/NaT si hay errores='coerce'.
                df_modificado[columna] = pd.to_numeric(df_modificado[columna], errors='coerce').astype('int64', errors='ignore')
            else:
                print(f"Advertencia: Tipo numérico '{tipo_numerico}' no reconocido para la columna '{columna}'.")
                continue


            print(f"Columna '{columna}' convertida a tipo '{tipo_numerico}'.")

        except Exception as e:
            print(f"Ocurrió un error inesperado al procesar la columna '{columna}': {e}")

    print(f"\nConversión a formato numérico completada para las columnas {columnas} del dataframe '{nombre}'.")

    return df_modificado


def reporte_calidad_global(DFs, nombres):
    """
    Crea un resumen de calidad de varios DataFrames.

    Parámetros:
      DFs: lista de DataFrames
      nombres: lista de nombres correspondientes a los DataFrames
    ADVERTENCIA: son pares ordenados los DFs y nombres. Deben tener igual orden y cantidad de elementos.

    Retorna:
      muestra un reporte de calidad de los DataFrames solo en pantalla.
    """
    resumen = []
    for df, nombre in zip(DFs, nombres):
        nulos = df.isna().sum().sum()                             # Total de valores nulos, no por columnas sino total, por eso el doble sum
        duplicados = df.duplicated(keep=False).sum()              # Total de filas duplicadas
        columnas = len(df.columns)                                # Cantidad de columnas
        filas = len(df)                                           # Cantidad de registros
        memoria = df.memory_usage(deep=True).sum() / 1024 / 1024  # Tamaño en MB

        resumen.append({
            "DataFrame": nombre,
            "Filas": filas,
            "Columnas": columnas,
            "Nulos totales": nulos,
            "Duplicados": duplicados,
            "Memoria": memoria
        })

    reporte = pd.DataFrame(resumen)
    display(reporte)


def categorizar_por_cuantiles(DF, columna, nombre_columna_nueva, nombre="", num_cuantiles=4, etiquetas=None):
    """
    Categoriza una columna numérica de un DataFrame en cuantiles.

    Args:
      DF: El DataFrame a modificar.
      columna: El nombre de la columna numérica a categorizar.
      nombre_columna_nueva: El nombre de la nueva columna con las categorías.
      nombre: Un nombre para el DataFrame (opcional).
      num_cuantiles: El número de cuantiles a utilizar (por defecto 4).
      etiquetas: Una lista de etiquetas para los cuantiles (opcional).
                 Si es None, se usarán etiquetas numéricas (0, 1, 2, ...).

    Returns:
      El DataFrame modificado con la nueva columna de categorías.
    """
    if DF is None:
        print("Error: El DataFrame no ha sido definido o es None.")
        return None

    if columna not in DF.columns:
        print(f"Error: La columna '{columna}' no existe en el DataFrame '{nombre}'.")
        return DF

    df_modificado = DF.copy()

    try:
        # Calcular los cuantiles
        cuantiles = df_modificado[columna].quantile(np.linspace(0, 1, num_cuantiles + 1)).tolist()

        # Asegurar que las etiquetas coincidan con el número de intervalos
        if etiquetas is not None and len(etiquetas) != num_cuantiles:
            print(f"Advertencia: El número de etiquetas ({len(etiquetas)}) no coincide con el número de cuantiles ({num_cuantiles}). Se usarán etiquetas numéricas.")
            etiquetas = None

        # Categorizar la columna
        df_modificado[nombre_columna_nueva] = pd.cut(df_modificado[columna], bins=cuantiles, labels=etiquetas, include_lowest=True)

        # Convertir a tipo category
        df_modificado[nombre_columna_nueva] = df_modificado[nombre_columna_nueva].astype('category')

        print(f"\nColumna '{columna}' del dataframe '{nombre}' categorizada por cuantiles en '{nombre_columna_nueva}'.")

    except Exception as e:
        print(f"Ocurrió un error al categorizar la columna '{columna}' por cuantiles: {e}")
        return DF

    return df_modificado

# C) Carga de los datos de entrada como DataFrames.

### Método de carga seleccionado.
> Evalué que método de carga de archivos usar en función de que el notebook en cuestión debe ser compartido con terceros que deben poder, no solo acceder al mismo, sino también a los archivos utilizados para la captura de datos de entrada.<br><br>
Para una experiencia mas fluida, menos técnica, mas clara y elegante, descarté la opción de hacerlo montando vía código la carpeta de Google Drive y preferí utilizar la carga mediante el uso de los ID de Google de los archivos de entrada necesarios.<br><br>
La solución propuesta, se enfoca en poder compartir con el tercero receptor del notebook solo un link a éste o un repositorio del mismo y no a toda la carpeta del proyecto, logrando mantener mayor control de la información compartida, facilitando el uso del mismo a usuarios menos técnicos y despejando el codigo de intrincadas rutas de acceso que desvían la atención de lo importante.<br><br>
Si bien en el caso particular no se esperan cambios de escala ni modificaciones constantes en el origen de los datos, se crean "constantes" en el código con los ID de cada archivo de origen para facilitar su manipulación y mantenimiento futuro sin tener que abordar múltiples líneas o secciones.
<br><br>
En el caso particular de acceder con un link de Google Drive y no desde un archivo .ipynb, se recomienda hacerlo en el navegador Chrome para evitar el recorte de funcionalidades. (obviamente la opción de guardar si permanecerá desactivada pero se podrá operar libremente y probar el notebook en cuestión.
<br><br>
Si por algún motivo no logra conexión con los id de Google Drive proporcionados y conoce los mismos pueden cargarse en el primer espacio de código como constantes.

### Código.

In [41]:
# IDs de G Drive de los archivos (.csv) de Base o entrada
ID_CSV_CLIENTES = '1O994MVX3EWF_fT0eYPGkL1KwwzUN7AYB'
ID_CSV_MARKETING = '1fU048eCW6MrHrYx_R3sGqCqkn-x0JHu1'
ID_CSV_VENTAS = '1psFzI9x9QSiRRAK2lJBwYb8FbnGJybJI'


In [42]:
# Generación del DataFrame de CLIENTES (Datos Crudos)
df_Clientes = pd.read_csv(generar_url_csv_gdrive(ID_CSV_CLIENTES))

In [43]:
# Generación del DataFrame de MARKETING (Datos Crudos)
df_Marketing = pd.read_csv(generar_url_csv_gdrive(ID_CSV_MARKETING))

In [44]:
# Generación del DataFrame de VENTAS (Datos Crudos)
df_Ventas = pd.read_csv(generar_url_csv_gdrive(ID_CSV_VENTAS))

# D) Exploración inicial de los datos.

## Visualización de los datos contenidos.

### Código.

In [45]:
# Se utiliza una función genérica para hacer una exploración inicial basada solo en la observacion.
# Función utilizada: "realizar_exploracion_inicial_ejemplo_visual".
realizar_exploracion_inicial_ejemplo_visual(DF=df_Clientes, nombre="df_Clientes", cantidad_lineas=8)



 'df_Clientes' Primeras filas:

   id_cliente               nombre  edad                 ciudad  ingresos
0           1      Aloysia Screase    44          Mar del Plata  42294.68
1           2  Kristina Scaplehorn    25                Posadas  24735.04
2           3       Filip Castagne    50            Resistencia  35744.85
3           4          Liuka Luard    39           Bahía Blanca  27647.96
4           5        Dore Cockshtt    28                Rosario  28245.65
5           6        Patrick Earle    34  San Miguel de Tucumán  62763.31
6           7           Etan Deeth    35            Resistencia  37489.71
7           8       Booth Bielfelt    40                Córdoba  35255.94


 'df_Clientes' Filas de ejemplo -al azar-:

     id_cliente             nombre  edad        ciudad  ingresos
22           23   Audrye Fairbeard    40      Santa Fe  46734.23
370         371     Yoshi Baysting    29  Bahía Blanca  25726.65
460         461          Pepi Bart    33  Buenos Aires  240

In [46]:
# Se utiliza una función genérica para hacer una exploración inicial basada solo en la observación.
# Función utilizada: "realizar_exploracion_inicial_ejemplo_visual".
realizar_exploracion_inicial_ejemplo_visual(DF=df_Marketing, nombre="df_Marketing", cantidad_lineas=8)



 'df_Marketing' Primeras filas:

   id_campanha          producto  canal  costo fecha_inicio   fecha_fin
0           74   Adorno de pared     TV   4.81   20/03/2024  03/05/2024
1           12            Tablet   RRSS   3.40   26/03/2024  13/05/2024
2           32   Lámpara de mesa  Email   5.54   28/03/2024  20/04/2024
3           21        Smartphone   RRSS   6.37   29/03/2024  16/05/2024
4           58          Alfombra  Email   4.25   31/03/2024  05/05/2024
5           85        SmartWatch     TV   5.07   01/04/2024  05/05/2024
6           36  Plancha de vapor  Email   5.41   02/04/2024  01/06/2024
7           57          Batidora  Email   4.48   10/04/2024  08/06/2024


 'df_Marketing' Filas de ejemplo -al azar-:

    id_campanha           producto  canal  costo fecha_inicio   fecha_fin
3            21         Smartphone   RRSS   6.37   29/03/2024  16/05/2024
0            74    Adorno de pared     TV   4.81   20/03/2024  03/05/2024
73           48     Cámara digital  Email   6.84

In [47]:
# Se utiliza una función genérica para hacer una exploración inicial basada solo en la observación.
# Función utilizada: "realizar_exploracion_inicial_ejemplo_visual".
realizar_exploracion_inicial_ejemplo_visual(DF=df_Ventas, nombre="df_Ventas", cantidad_lineas=8)




 'df_Ventas' Primeras filas:

   id_venta           producto   precio  cantidad fecha_venta  \
0       792  Cuadro decorativo   $69.94       5.0  02/01/2024   
1       811    Lámpara de mesa  $105.10       5.0  02/01/2024   
2      1156           Secadora   $97.96       3.0  02/01/2024   
3      1372           Heladera  $114.35       8.0  02/01/2024   
4      1546           Secadora  $106.21       4.0  02/01/2024   
5      1697    Horno eléctrico   $35.35       9.0  02/01/2024   
6      1710   Plancha de vapor   $65.43       2.0  02/01/2024   
7      2959          Proyector   $88.17       9.0  02/01/2024   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos  
4  Electrodomésticos  
5  Electrodomésticos  
6  Electrodomésticos  
7        Electrónica  


 'df_Ventas' Filas de ejemplo -al azar-:

      id_venta             producto   precio  cantidad fecha_venta  \
1530       383    Espejo decorativo  $119.48       7.0  24/06

### Conclusiones preliminares (paso 1).
>*Las conclusiones se acumularán recursivamente durante los pasos de exploración de los datos, en caso de querer revisar directamente la última versión de las mismas dirigirse a "Conclusiones Finales" en el último punto de la sección.*

#### dF_Clientes:

*   Orden Ascendente columna "id_cliente".
*   Columna "id_cliente" debería ser Int(16) si no se espera una escalada importante de cantidad.
*   Columna "nombre" deberia limpiarse y normalizarse dado que contiene caracteres especiales y espacios. Se observan ' intermedios.
*   Columna "edad" deberia ser Int(8).
*   Columna "ciudad" deberia limpiarse y normalizarse dado que contiene caracteres especiales y espacios.
*   Columna "ingresos" deberia ser float(64), El formato parece correcto tiene "." como separación decimal. VER: No queda claro a que se refiere con ingresos, es un sum(cliente.ventas)?

#### df_Marketing:

*   Orden Ascendente columna "fecha_inicio".  Fecha inicio 20/03/2024 - Fecha fin 21/01/2025.
*   Columna "id_campanha" debería ser Int(16) si no se espera una escalada importante de cantidad.
*   Columna "producto" debería limpiarse y normalizarse dado que contiene caracteres especiales y espacios. Se observan ' intermedios.
*   Columna "canal" debería ser "category " si encuadra en analisis detallado.
*   Columna "costo" debería ser float(64), El formato parece correcto tiene "." como separación decimal. VER: Parece expresado en "Miles de $"
*   Columna "fecha_inicio" deberia ser datetime64. Esta en formato "dd/mm/aaaa".
*   Columna "fecha_fin" deberia ser datetime64. Esta en formato "dd/mm/aaaa".
*   Pareciera útil una nueva columna "duracion_campanha" con formato timedelta.
*   Pareciera útil una nueva columna "categoria" con la categorizacion que se usa para Producto (Ver de "df_Ventas" para análisis mas globales). Tambien puede generarse una tabla complementaria.
*   El orden actual es el mas conveniente?



#### df_Ventas:

*   Orden Ascendente columna "fecha_venta". Fecha inicio 02/01/2024 - Fecha fin 30/12/2024.
*   Posee en común la columna "Producto" con "df_Marketing"
*   Columna "id_venta" debería ser Int(16) si no se espera una escalada importante de cantidad.
*   Columna "producto" debería limpiarse y normalizarse dado que contiene caracteres especiales y espacios. Se observan ' intermedios.
*   Columna "precio" debería ser float(64), El formato tiene un "\$" al inicio y "." como separación decimal. VER: Quitar caracter "\$" y confirmar que es el precio unitario.
*   Columna "cantidad" debería ser Int(16) o float(16) si posee cantidades decimales y no se prevee una escalada importante de cantidad.
*   Columna "fecha_venta" debería ser datetime64. Esta en formato "dd/mm/aaaa".
*   Columna "categoria" debería ser category. Esta en formato "dd/mm/aaaa". debería limpiarse y normalizarse dado que contiene caracteres especiales y espacios.
*   Pareciera útil una nueva columna "precio_venta" con formato float(64).
*   El orden actual es el mas conveniente?
*   Pareciera útil una nueva columna "id_cliente" para unir la indormacion con "df_Clientes". -Por ahora dato de existencia desconocida-
*   En teoría "categoria" es una categorización de Producto.

#### Observaciones Generales

*   Los datos parecen medianamente bien estructurados y de fácil normalizción y limpieza.
*   Se recomienda buscar medios para vincular "df_Clientes" con "df_Ventas".

## Exploración profunda de los datos.

### Código.

In [48]:
# Se utiliza una función genérica para hacer una exploración inicial profunda basada en los datos del DataFrame.
# Función utilizada: "realizar_exploracion_inicial_profunda".
realizar_exploracion_inicial_profunda(DF=df_Clientes, nombre="df_Clientes")



 'df_Clientes' Analisis de Tamaño: (Filas, Columnas)

(567, 5)


 'df_Clientes' Tipos de Datos, Nulos, etc.:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 567 entries, 0 to 566
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id_cliente  567 non-null    int64  
 1   nombre      567 non-null    object 
 2   edad        567 non-null    int64  
 3   ciudad      567 non-null    object 
 4   ingresos    567 non-null    float64
dtypes: float64(1), int64(2), object(2)
memory usage: 22.3+ KB
None


 'df_Clientes' Informacion Estadistica del set:

        id_cliente       nombre        edad         ciudad      ingresos
count   567.000000          567  567.000000            567    567.000000
unique         NaN          567         NaN             12           NaN
top            NaN  Micah Matis         NaN  Mar del Plata           NaN
freq           NaN            1         NaN             63           NaN
mean    284.0

In [49]:
# Se utiliza una función genérica para hacer una exploración inicial profunda basada en los datos del DataFrame.
# Función utilizada: "realizar_exploracion_inicial_profunda".
realizar_exploracion_inicial_profunda(DF=df_Marketing, nombre="df_Marketing")



 'df_Marketing' Analisis de Tamaño: (Filas, Columnas)

(90, 6)


 'df_Marketing' Tipos de Datos, Nulos, etc.:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id_campanha   90 non-null     int64  
 1   producto      90 non-null     object 
 2   canal         90 non-null     object 
 3   costo         90 non-null     float64
 4   fecha_inicio  90 non-null     object 
 5   fecha_fin     90 non-null     object 
dtypes: float64(1), int64(1), object(4)
memory usage: 4.3+ KB
None


 'df_Marketing' Informacion Estadistica del set:

        id_campanha         producto canal      costo fecha_inicio   fecha_fin
count     90.000000               90    90  90.000000           90          90
unique          NaN               30     3        NaN           78          78
top             NaN  Adorno de pared    TV        NaN   25/09/2024  05/05/2024
freq  

In [50]:
# Se utiliza una función genérica para hacer una exploración inicial profunda basada en los datos del DataFrame.
# Funci+on utilizada: "realizar_exploracion_inicial_profunda".
realizar_exploracion_inicial_profunda(DF=df_Ventas, nombre="df_Ventas")



 'df_Ventas' Analisis de Tamaño: (Filas, Columnas)

(3035, 6)


 'df_Ventas' Tipos de Datos, Nulos, etc.:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3035 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3035 non-null   int64  
 1   producto     3035 non-null   object 
 2   precio       3033 non-null   object 
 3   cantidad     3033 non-null   float64
 4   fecha_venta  3035 non-null   object 
 5   categoria    3035 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 142.4+ KB
None


 'df_Ventas' Informacion Estadistica del set:

           id_venta         producto  precio     cantidad fecha_venta  \
count   3035.000000             3035    3033  3033.000000        3035   
unique          NaN               30    2590          NaN         364   
top             NaN  Lámpara de mesa  $76.32          NaN  06/04/2024   
freq            NaN              181   

### Conclusiones preliminares (paso 2).
>*Las conclusiones se acumularán recursivamente durante los pasos de exploración de los datos, en caso de querer revisar directamente la última versión de las mismas dirigirse a "Conclusiones Finales" en el último punto de la sección.*

#### dF_Clientes:

*   Cantidad de registros 567.
*   Orden Ascendente columna "id_cliente". Al parecer del 1 al 567 de uno en uno => Identificador único de Clientes.
*   No se registran nulos en ninguna de las columnas.
*   Columna "id_cliente" es int(64) debería ser int(16) si no se espera una escalada importante de cantidad.
*   Columna "nombre" debería limpiarse y normalizarse dado que contiene caracteres especiales y espacios. Se observan ' intermedios. Al parecer hay 567 registros distintos por lo que no parece registrar duplicados, aunque podría pasar la existencia de homónimos sin que sea un error.
*   Columna "edad" es int(64)y debería ser Int(8). Los valores contenidos tienen lógica y se encuentran entre 20-81.
*   Columna "ciudad" debería limpiarse y normalizarse dado que contiene caracteres especiales y espacios. Al parecer debería ser "category" posee 12 ocurrencias distintas.
*   Columna "ingresos" debería ser float(64), El formato parece correcto tiene "." como separación decimal. VER: No queda claro a que se refiere con ingresos, es un sum(cliente.ventas)?
*   Pareciera útil una nueva columna "rango_edad" para segmentar los estudios por rango (tipo de dato "category"). Aparenta distribución normal. Entre los 30 y 43 parece concentrarse el 50% de las ocurrencias.
*	Pareciera útil una nueva columna "rango_ingresos" para segmentar los estudios por rango (tipo de dato "category"). Aparenta distribución normal. Entre los 26k y 42.5k parece concentrarse el 50% de las ocurrencias.
*   De ser posible sumar ingresos de "df_Clientes" y comparar contra la nueva "precio_venta" adicional a "df_Ventas".
*   Para futuros Indexados se recomienda "id_cliente" como principal y como secundarios "ciudad", "ingresos" y "rango_edad

#### df_Marketing:

*   Cantidad de registros 90.
*   Orden Ascendente columna "fecha_inicio".  Fecha inicio 20/03/2024 - Fecha fin 21/01/2025.
*   No se registran nulos en ninguna de las columnas.
*   Columna "id_campanha" es int(64) y debería ser Int(16) si no se espera una escalada importante de cantidad. Al parecer del 1 al 90 de uno en uno => Identificador único de Campanhas.
*   Columna "producto" debería limpiarse y normalizarse dado que contiene caracteres especiales y espacios. Se observan ' intermedios. La cantidad de Productos es de 30 / 90 movimientos el uso del formato "category" no parece indispensable aunque puede que mejore el uso de memoria si se une con "dF_Ventas". En "dF_Ventas" la cantidad de productos también es de 30 por lo que todos los productos vendidos efectuaron al menos 1 campaña de marketing.
*   Columna "canal" debería ser "category" son 3 categorias / 90 registros.
*   Columna "costo" debería ser float(64), El formato parece correcto tiene "." como separación decimal. VER: Parece expresado en "Miles de $"
*   Columna "fecha_inicio" debería ser datetime64. Esta en formato "dd/mm/aaaa".
*   Columna "fecha_fin" debería ser datetime64. Esta en formato "dd/mm/aaaa".
*   Pareciera útil una nueva columna "mes_inicio", "mes_fin" int(8) para efectuar cruces temporales categorizados con otros DFs. El volumen de datos justifica hacerlo en una columna y priorizar legibilidad a hacerlo "internamente" en el mismo paso en que se utilice.
*   Pareciera útil una nueva columna "rango_costo" para segmentar los estudios por rango (tipo de dato "category"). Aparenta Distribucion Normal. Entre los 4.37k y 5.56k parece concentrarse el 50% de las ocurrencias.
*   Pareciera útil una nueva columna "duracion_campanha" con formato timedelta o int si es en dias, ver. Incluir control de delta cero o negativo para detectar errores en el crudo. De proceder puede crearse una nueva por rango, "rango_duracion_campanha"
*   Pareciera útil una nueva columna "categoria" con la categorización que se usa para el Producto (Ver de "df_Ventas" para análisis mas globales). También puede generarse una tabla complementaria.
*   El orden actual es el mas conveniente? - Entiendo que producto, seria la indexacion principal, "fecha_inicio", "rango_costo" y "rango_duracion_campanha" secundarios.
*   Se podrían agregar datos en una nueva tabla comparando con la información de "df_Ventas" para medir cantidad de operaciones,  categorizando frecuencia, estacionalidad, presunción de continuar activo, etc.
*   Se podrían agregar datos en una nueva tabla comparando con la información de "df_Ventas" para analizar la relación cliente/producto para direccionar futuras campañas de marketing.


#### df_Ventas:

*   Cantidad de registros 3035.
*   Orden Ascendente columna "fecha_venta". Fecha inicio 02/01/2024 - Fecha fin 30/12/2024.
*   No se registran nulos en ninguna de las columnas.
*   Posee en común la columna "Producto" con "df_Marketing"
*   Columna "id_venta" debería ser Int(16) si no se espera una escalada importante de cantidad. El valor mínimo es 1 y el máximo es 3000, por lo que no es un identificador único, revisar y entender por que.
*   Columna "producto" debería limpiarse y normalizarse dado que contiene caracteres especiales y espacios. Se observan ' intermedios. Incluye 30 productos distintos.
*   Columna "precio" debería ser float(64), El formato tiene un "\$" al inicio y "." como separación decimal. VER: Quitar caracter "\$" y confirmar que es el precio unitario.
*   Columna "cantidad" debería ser Int(16) o float(16) si posee cantidades decimales y no se espera una escalada importante de cantidad.
*   Columna "fecha_venta" debería ser datetime64. Esta en formato "dd/mm/aaaa".
*   Columna "categoria" debería ser category. Esta en formato "dd/mm/aaaa". debería limpiarse y normalizarse dado que contiene caracteres especiales y espacios.
*   Pareciera útil una nueva columna "mes_venta" int(8) o Datatime? para efectuar cruces temporales categorizados con otros DataFrames. El volumen de datos justifica hacerlo en una columna y priorizar legibilidad a hacerlo "internamente" en el mismo paso en que se utilice.
*   Pareciera útil una nueva columna "importe_bruto_venta" con formato float(64).
*   Pareciera útil una nueva columna "rango_importe_bruto_venta" para segmentar los estudios por rango (tipo de dato "category"). Verificar distribución una vez creada "importe_bruto_venta".
*   El orden actual es el mas conveniente?
*   Pareciera útil una nueva columna "id_cliente" para unir la información con "df_Clientes". -Por ahora dato de existencia desconocida-
*   En teoría "categoria" es una categorización de Producto. Hay 3 Categorias.

#### Observaciones Generales

*   Los datos parecen medianamente bien estructurados y de fácil normalizción y limpieza.
*   Con los datos actuales se puede analizar las relaciones "precio/cantidad" y "precio/tiempo" para comprender si hay bonificaciones por cantidad, volatilidad en los precios, cambios estacionarios de precios, etc.
*   Se recomienda buscar medios para vincular "df_Clientes" con "df_Ventas".
*   Sería interesante poseer información de costos por unidad de producto para poder hacer análisis de "utilidad/producto"

# E) Validación de rangos y valores.
> Durante la exploración efectuada en las secciones precedentes, no se encontraron valores disonantes ni comportamientos inesperados que motiven un análisis mas profundo sobre los rangos y valores que se observaron en los distintos DataFrames analizados.

# F) Detección y tratamiento de valores nulos.

> Por los análisis preliminares globales efectuados no parece observarse valores nulos, igualmente se procede a revisar los datos mediante la función general "conteo_nulos_por_columna", que revisa columna por columna de cada DataFrame.

### Código.

In [51]:
# Se utiliza una función genérica para detectar nulos en las columnas del DataFrame.
# Función utilizada: "conteo_nulos_por_columna".
contar_nulos_por_columna(DF=df_Clientes, nombre="df_Clientes")


Conteo de valores nulos por columna del dataframe 'df_Clientes':

Columna 'id_cliente': 0 valores nulos
Columna 'nombre': 0 valores nulos
Columna 'edad': 0 valores nulos
Columna 'ciudad': 0 valores nulos
Columna 'ingresos': 0 valores nulos


In [52]:
# Se utiliza una función genérica para detectar nulos en las columnas del DataFrame.
# Función utilizada: "conteo_nulos_por_columna".
contar_nulos_por_columna(DF=df_Marketing, nombre="df_Marketing")


Conteo de valores nulos por columna del dataframe 'df_Marketing':

Columna 'id_campanha': 0 valores nulos
Columna 'producto': 0 valores nulos
Columna 'canal': 0 valores nulos
Columna 'costo': 0 valores nulos
Columna 'fecha_inicio': 0 valores nulos
Columna 'fecha_fin': 0 valores nulos


In [53]:
# Se utiliza una función genérica para detectar nulos en las columnas del DataFrame.
# Función utilizada: "conteo_nulos_por_columna".
contar_nulos_por_columna(DF=df_Ventas, nombre="df_Ventas")


Conteo de valores nulos por columna del dataframe 'df_Ventas':

Columna 'id_venta': 0 valores nulos
Columna 'producto': 0 valores nulos
Columna 'precio': 2 valores nulos
Columna 'cantidad': 2 valores nulos
Columna 'fecha_venta': 0 valores nulos
Columna 'categoria': 0 valores nulos


In [54]:
mostrar_nulos_de_una_columna(DF=df_Ventas, nombre="df_Ventas", columna='precio')
mostrar_nulos_de_una_columna(DF=df_Ventas, nombre="df_Ventas", columna='cantidad')


Filas con valores nulos en la columna 'precio' del dataframe 'df_Ventas':

     id_venta               producto precio  cantidad fecha_venta    categoria
136       627  Elementos de cerámica    NaN       NaN  17/01/2024   Decoración
139      2171    Parlantes Bluetooth    NaN       NaN  17/01/2024  Electrónica

Filas con valores nulos en la columna 'cantidad' del dataframe 'df_Ventas':

     id_venta               producto precio  cantidad fecha_venta    categoria
136       627  Elementos de cerámica    NaN       NaN  17/01/2024   Decoración
139      2171    Parlantes Bluetooth    NaN       NaN  17/01/2024  Electrónica


>#### Se detectaron dos filas con valores nulos tanto en 'cantidad' como en 'precio'. Para determinar el procedimiento a seguir se hacen algunos análisis específicos para entender el problema y responder con la mejor solución posible.

In [55]:
# Se Analizaron todas las ocurrencias de los productos que poseen el error en el DataFrame.

# Filtrar 'df_Ventas por producto' - 'Elementos de cerámica'.
precios_ventas_elementos_de_ceramica = df_Ventas[df_Ventas['producto'] == 'Elementos de cerámica']

# Mostrar las filas filtradas.
display(precios_ventas_elementos_de_ceramica)

id_venta               producto   precio  cantidad fecha_venta  \
36          17  Elementos de cerámica  $108.48      11.0  07/01/2024   
51         687  Elementos de cerámica   $27.68       8.0  08/01/2024   
71         817  Elementos de cerámica  $106.31       1.0  10/01/2024   
95         507  Elementos de cerámica  $111.10       1.0  13/01/2024   
97         987  Elementos de cerámica  $102.62       3.0  13/01/2024   
...        ...                    ...      ...       ...         ...   
2892       427  Elementos de cerámica   $60.13       6.0  14/12/2024   
2904       567  Elementos de cerámica   $97.22       8.0  15/12/2024   
2961       917  Elementos de cerámica   $52.65       8.0  22/12/2024   
2972       497  Elementos de cerámica  $124.16       4.0  23/12/2024   
3011       167  Elementos de cerámica  $104.96       6.0  28/12/2024   

       categoria  
36    Decoración  
51    Decoración  
71    Decoración  
95    Decoración  
97    Decoración  
...          ...  
2892  Decoración  
2904  Decoración  
2961  Decoración  
2972  Decoración  
3011  Decoración  

[100 rows x 6 columns]

In [56]:
# Filtrar 'df_Ventas por producto' - 'Parlantes Bluetooth'.
precios_ventas_parlantes_bluetooth = df_Ventas[df_Ventas['producto'] == 'Parlantes Bluetooth']

# Mostrar las filas filtradas.
display(precios_ventas_parlantes_bluetooth)

id_venta             producto  precio  cantidad fecha_venta    categoria
56        2501  Parlantes Bluetooth  $87.78       1.0  08/01/2024  Electrónica
87        2761  Parlantes Bluetooth  $29.89       9.0  11/01/2024  Electrónica
139       2171  Parlantes Bluetooth     NaN       NaN  17/01/2024  Electrónica
151       2431  Parlantes Bluetooth  $69.40       1.0  18/01/2024  Electrónica
162       2231  Parlantes Bluetooth  $42.98       7.0  19/01/2024  Electrónica
...        ...                  ...     ...       ...         ...          ...
2928      2921  Parlantes Bluetooth  $97.14       8.0  16/12/2024  Electrónica
2954      2451  Parlantes Bluetooth  $42.08       2.0  20/12/2024  Electrónica
2994      2511  Parlantes Bluetooth  $52.04       5.0  25/12/2024  Electrónica
3004      2961  Parlantes Bluetooth  $29.97       3.0  26/12/2024  Electrónica
3009      2951  Parlantes Bluetooth  $89.89       6.0  27/12/2024  Electrónica

[99 rows x 6 columns]

In [57]:
# Como ambos nulos se evidenciaron en una única fecha, se listaron todos los
# movimientos de la misma por si se dectectan otras anomalías ocurridas o se
# logra inferir la causa del problema.

# Filtrar df_Ventas por producto 'Parlantes Bluetooth'
ventas_17_01_2024 = df_Ventas[df_Ventas['fecha_venta'] == '17/01/2024']

# Mostrar las filas filtradas
display(ventas_17_01_2024)

id_venta               producto   precio  cantidad fecha_venta  \
133       131        Lámpara de mesa   $58.32      10.0  17/01/2024   
134       178      Rincón de plantas   $81.50       1.0  17/01/2024   
135       368      Rincón de plantas   $87.40       9.0  17/01/2024   
136       627  Elementos de cerámica      NaN       NaN  17/01/2024   
137       684      Jarrón decorativo  $106.89       6.0  17/01/2024   
138      1750       Plancha de vapor   $94.53       2.0  17/01/2024   
139      2171    Parlantes Bluetooth      NaN       NaN  17/01/2024   
140      2284                 Tablet   $69.35       4.0  17/01/2024   
141      2402              Televisor  $117.53       9.0  17/01/2024   
142      2698             SmartWatch   $76.47       7.0  17/01/2024   

             categoria  
133         Decoración  
134         Decoración  
135         Decoración  
136         Decoración  
137         Decoración  
138  Electrodomésticos  
139        Electrónica  
140        Electrónica  
141        Electrónica  
142        Electrónica

### Conclusiones preliminares (paso 3).
> Se observa que no hay una causa determinante observable para la aparición de los valores nulos, y que los mismos corresponden a únicamente dos (2) registros o filas del DataFrame en cuestión. Cuantitativamente, sobre un total de 3035 casos, representan solo el 7/10000 y dentro de su sub grupo 'producto', representan 1/100.
>Por otra parte, al analizar cada 'producto' en particular se desprende que dicha columna no detalla productos únicos, sino que se trata de un agrupamiento categórico de varios modelos/artículos de un grupo afín, se llega a esta conclusión basada en la gran dispersión del precio unitario en las distintas ventas lo que implica una compleja identificación del producto específico en los registros con problemas.
> Dado la notoria insignificancia del dato y la imposibilidad de asignar un valor razonable para remplazar los valores nulos, se sugiere **eliminar las filas en su totalidad**.
>Para un mayor control, en esta sección se solamente se apuntan los cambios sugeridos a realizar y finalmente, luego de todos los análisis preliminares y de exploración, se procederá a realizar todos los ajustes propuestos.

# G) Detección y tratamiento de duplicados.

### Código.

In [58]:
# Se utiliza una función genérica para detectar duplicados de fila completa del DataFrame.
# Función utilizada: "detectar_duplicados".
detectar_duplicados(DF=df_Clientes, nombre="df_Clientes")


Conteo de valores duplicados en el dataframe 'df_Clientes':

Cantidad de filas duplicadas: 0


In [59]:
# Se utiliza una función genérica para detectar duplicados de fila completa del DataFrame.
# Función utilizada: "detectar_duplicados".
detectar_duplicados(DF=df_Marketing, nombre="df_Marketing")


Conteo de valores duplicados en el dataframe 'df_Marketing':

Cantidad de filas duplicadas: 0


In [60]:
# Se utiliza una función genérica para detectar duplicados de fila completa del DataFrame.
# Función utilizada: "detectar_duplicados".
# Como en una primera ejecución se encontraron duplicados, se agregó el dato de
# la columna que es "primary Key" (en este caso es una sola y no un conjunto)
detectar_duplicados(DF=df_Ventas, nombre="df_Ventas", columnas_pk=["id_venta"])


Conteo de valores duplicados en el dataframe 'df_Ventas' por ['id_venta']:

Cantidad de filas duplicadas: 35

Filas duplicadas en el dataframe 'df_Ventas' por ['id_venta']:



id_venta  cantidad_duplicados
0         56                    2
1        145                    2
2        204                    2
3        213                    2
4        300                    2
5        345                    2
6        398                    2
7        421                    2
8        424                    2
9        439                    2
10       509                    2
11       821                    2
12       906                    2
13       959                    2
14      1376                    2
15      1381                    2
16      1642                    2
17      1776                    2
18      1784                    2
19      1868                    2
20      1956                    2
21      1974                    2
22      2080                    2
23      2085                    2
24      2096                    2
25      2108                    2
26      2169                    2
27      2356                    2
28      2365                    2
29      2373                    2
30      2486                    2
31      2506                    2
32      2545                    2
33      2705                    2
34      2778                    2

### Conclusiones preliminares (paso 4).
>Los DataFrames 'df_Clientes' y 'df_Marketing' no poseen duplicados, mientras que el 'df_Ventas' cuneta con 35 claves únicas repetidas (id_venta), lo que explica que la cantidad de registros totales de la columna mencionada sean 3035 y que posea solamente 3000 valores únicos -observado durante la exploración profunda-. Se recomienda eliminar las filas duplicadas en la etapa de transformación / limpieza de los datos.


# H) Limpieza y transformación de los DataFrames.

## a) Copia / reserva de los DataFrames originales.
>Para posibilitar comparaciones o reversiones posteriores, se realiza una copia de los DataFrames originales para operar las transformaciones sobre estos.




In [61]:
# Se crean copias de los DataFrames originales
df_Clientes_v2 = df_Clientes.copy()
df_Marketing_v2 = df_Marketing.copy()
df_Ventas_v2 = df_Ventas.copy()

print("Copias de los DataFrames originales creadas en su 'Version 2':")
print(f"df_Clientes_v2: {df_Clientes_v2.shape}")
print(f"df_Marketing_v2: {df_Marketing_v2.shape}")
print(f"df_Ventas_v2: {df_Ventas_v2.shape}")

Copias de los DataFrames originales creadas en su 'Version 2':
df_Clientes_v2: (567, 5)
df_Marketing_v2: (90, 6)
df_Ventas_v2: (3035, 6)


## b) Eliminación de duplicados.
> Se procede a eliminar los duplicados detectados en el punto G).


### Código y verificación.

In [62]:
# Se utiliza una función genérica para eliminar duplicados de fila completa del DataFrame.
# Función utilizada: "eliminar_duplicados".
print(f"df_Ventas_v2 ANTES: {df_Ventas_v2.shape}\n ")
df_Ventas_v2 = eliminar_duplicados(DF=df_Ventas_v2, nombre="df_Ventas_v2", columnas_pk=["id_venta"], tipo_de_eliminacion='first')
print(f"\n df_Ventas_v2 DESPUES: {df_Ventas_v2.shape}")
# Verificacion:
print("\n\n VERIFICACION:")
detectar_duplicados(DF=df_Ventas_v2, nombre="df_Ventas_v2", columnas_pk=["id_venta"])

df_Ventas_v2 ANTES: (3035, 6)
 

Duplicados eliminados del dataframe 'df_Ventas_v2' por ['id_venta']:
Cantidad de filas eliminadas: 35

 df_Ventas_v2 DESPUES: (3000, 6)


 VERIFICACION:

Conteo de valores duplicados en el dataframe 'df_Ventas_v2' por ['id_venta']:

Cantidad de filas duplicadas: 0


## c) Eliminación / reemplazo de valores nulos.
> Se procede a eliminar las filas con valores nulos según lo detallado en el punto F).

### Código y verificación.

In [63]:
# Se utiliza una función genérica para intervenir los valores nulos de una columna del DataFrame.
# Función utilizada: "intervenir_nulos_de_una_columna".
print(f"df_Ventas_v2 ANTES: {df_Ventas_v2.shape}")
df_Ventas_v2 = intervenir_nulos_de_una_columna(DF=df_Ventas_v2, nombre= 'df_Ventas_v2', columna='precio', tipo_operacion='eliminar')
# Dado que las filas con nulos en la columna 'cantidad' son las mismas con nulos
# en 'precio', no es necesario hacer una segunda eliminación.
print(f"\n df_Ventas_v2 DESPUES: {df_Ventas_v2.shape}")
# Verificacion:
print("\n\n VERIFICACION:")
contar_nulos_por_columna(DF=df_Ventas_v2, nombre="df_Ventas_v2")

df_Ventas_v2 ANTES: (3000, 6)

Interviniendo valores nulos en la columna 'precio' del dataframe 'df_Ventas_v2' con operación 'eliminar':
Se eliminaron las filas con valores nulos en la columna 'precio'.
Cantidad de valores nulos antes: 2, después: 0. Filas afectadas: 2

 df_Ventas_v2 DESPUES: (2998, 6)


 VERIFICACION:

Conteo de valores nulos por columna del dataframe 'df_Ventas_v2':

Columna 'id_venta': 0 valores nulos
Columna 'producto': 0 valores nulos
Columna 'precio': 0 valores nulos
Columna 'cantidad': 0 valores nulos
Columna 'fecha_venta': 0 valores nulos
Columna 'categoria': 0 valores nulos


## d) Normalización de Datos

*   Primero, se procede a normalizar todas las columnas de texto. Las columnas a modificar se seleccionan manualmente para un mayor control según los datos obtenidos en D). Se aplica una solución genérica pero personalizada para cada caso, explicitando el formato a "Texto-STR", quitando todos los comandos de "manejo de espacios", espacios en blanco innecesarios, se elije aplicar formato "Title"  o "UPPER" -mayuscula la primera letra, o todo en mayuscula- segun el caso y se decide modificar los caracteres propios soportados en UTF8 a estandares mas amigables en caso de tener que usar los datos en aplicaciones o soluciones que no los soportan (igualmente en caso de que se quiera, al momento de mostrar la información a terceros, puede volver a convertirse facilmente con la expresion regular inversa correspondiente).
*   Segundo, se procede ajustar el formato de las fechas para representarlas como datatime64, lo que posibilitará en el futuro realizar multiples operaciones con fechas y datatimes. (Más información de formatos de fecha/hora ver: [documentacion de Python](https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes))
*   Tercero, se ajustan los registros que deberian ser numéricos para mostrarse como tales.

### Código y verificación.

In [64]:
# Se utiliza una función genérica para normalizar el texto de las columnas de
# cada DataFrame según lo observado en el punto D).
# Función utilizada: "normalizar_texto".
print(f"df_Clientes_v2 ANTES: \n{df_Clientes_v2.head(8)}\n\n{df_Clientes_v2.dtypes}\n")
df_Clientes_v2 = normalizar_texto(
    DF = df_Clientes_v2,
    columnas = ['nombre', 'ciudad'],
    nombre = "df_Clientes_v2",
    eliminar_espacios_extra = True,
    eliminar_caracteres_especiales = True,
    tipo_conversion = 'title',
    mapeo_caracteres_especiales = None # Usa el Mapeo por defecto en remover_caracteres_especiales -quita UTF8-
  )
# No registra columnas con Fechas y los valores numéricos estan correctamente formateados.
print(f"\n df_Clientes_v2 DESPUES: \n{df_Clientes_v2.head(8)}\n\n{df_Clientes_v2.dtypes}\n")

df_Clientes_v2 ANTES: 
   id_cliente               nombre  edad                 ciudad  ingresos
0           1      Aloysia Screase    44          Mar del Plata  42294.68
1           2  Kristina Scaplehorn    25                Posadas  24735.04
2           3       Filip Castagne    50            Resistencia  35744.85
3           4          Liuka Luard    39           Bahía Blanca  27647.96
4           5        Dore Cockshtt    28                Rosario  28245.65
5           6        Patrick Earle    34  San Miguel de Tucumán  62763.31
6           7           Etan Deeth    35            Resistencia  37489.71
7           8       Booth Bielfelt    40                Córdoba  35255.94

id_cliente      int64
nombre         object
edad            int64
ciudad         object
ingresos      float64
dtype: object


 Normalización de texto aplicada a las columnas ['nombre', 'ciudad'] del dataframe 'df_Clientes_v2'.

 df_Clientes_v2 DESPUES: 
   id_cliente               nombre  edad                

In [65]:
# Se utiliza una función genérica para normalizar el texto de las columnas de
# cada DataFrame según lo observado en el punto D).
# Función utilizada: "normalizar_texto".
print(f"df_Marketing_v2 ANTES: \n{df_Marketing_v2.head(8)}\n\n{df_Marketing_v2.dtypes}\n")
df_Marketing_v2 = normalizar_texto(
    DF = df_Marketing_v2,
    columnas = ['producto', 'canal'],
    nombre = "df_Marketing_v2",
    eliminar_espacios_extra = True,
    eliminar_caracteres_especiales = True,
    tipo_conversion = 'upper',
    mapeo_caracteres_especiales = None # Usa el Mapeo por defecto en remover_caracteres_especiales -quita UTF8-
  )
# Respecto a las fechas, se utiliza una función genérica para normalizarlas.
# Función utilizada: "convertir_formato_fecha".
df_Marketing_v2 = convertir_formato_fecha(
     DF = df_Marketing_v2,
    columnas = ['fecha_inicio', 'fecha_fin'],
    nombre = "df_Marketing_v2",
    formato_original = "%d/%m/%Y"
   )
# Los valores numéricos estan correctamente formateados.(Ver ajustar titulo columna avisando que esta en miles de $)
print(f"\n df_Marketing_v2 DESPUES: \n{df_Marketing_v2.head(8)}\n\n{df_Marketing_v2.dtypes}\n")

df_Marketing_v2 ANTES: 
   id_campanha          producto  canal  costo fecha_inicio   fecha_fin
0           74   Adorno de pared     TV   4.81   20/03/2024  03/05/2024
1           12            Tablet   RRSS   3.40   26/03/2024  13/05/2024
2           32   Lámpara de mesa  Email   5.54   28/03/2024  20/04/2024
3           21        Smartphone   RRSS   6.37   29/03/2024  16/05/2024
4           58          Alfombra  Email   4.25   31/03/2024  05/05/2024
5           85        SmartWatch     TV   5.07   01/04/2024  05/05/2024
6           36  Plancha de vapor  Email   5.41   02/04/2024  01/06/2024
7           57          Batidora  Email   4.48   10/04/2024  08/06/2024

id_campanha       int64
producto         object
canal            object
costo           float64
fecha_inicio     object
fecha_fin        object
dtype: object


 Normalización de texto aplicada a las columnas ['producto', 'canal'] del dataframe 'df_Marketing_v2'.

Convirtiendo formato de fecha y tipo de datos a datetime64 en e

In [66]:
# Se utiliza una función genérica para normalizar el texto de las columnas de
# cada DataFrame según lo observado en el punto D).
# Función utilizada: "normalizar_texto".
print(f"df_Ventas_v2 ANTES: \n{df_Ventas_v2.head(8)}\n\n{df_Ventas_v2.dtypes}\n")
df_Ventas_v2 = normalizar_texto(
    DF = df_Ventas_v2,
    columnas = ['producto', 'categoria'],
    nombre = "df_Ventas_v2",
    eliminar_espacios_extra = True,
    eliminar_caracteres_especiales = True,
    tipo_conversion = 'upper',
    mapeo_caracteres_especiales = None # Usa el Mapeo por defecto en remover_caracteres_especiales -quita UTF8-
  )
# Respecto a las fechas, se utiliza una función genérica para normalizarlas.
# Función utilizada: "convertir_formato_fecha".
df_Ventas_v2 = convertir_formato_fecha(
     DF = df_Ventas_v2,
    columnas = ['fecha_venta'],
    nombre = "df_Marketing_v2",
    formato_original = "%d/%m/%Y"
   )
# Respecto a los datos numericos, se utiliza una función genérica para normalizarlos.
# Función utilizada: "convertir_formato_numerico".
# Se opta por dejar la cantidad como float dado que por la calidad general de los
# datos originales se nota que se dejó proparado para posibilitar decimales.
df_Ventas_v2 = convertir_formato_numerico(
    DF = df_Ventas_v2,
    columnas = ['precio', 'cantidad'],
    nombre = "df_Ventas_v2",
    tipo_numerico = 'float64'
)
print(f"\n df_Ventas_v2 DESPUES: \n{df_Ventas_v2.head(8)}\n\n{df_Ventas_v2.dtypes}\n")

df_Ventas_v2 ANTES: 
   id_venta           producto   precio  cantidad fecha_venta  \
0       792  Cuadro decorativo   $69.94       5.0  02/01/2024   
1       811    Lámpara de mesa  $105.10       5.0  02/01/2024   
2      1156           Secadora   $97.96       3.0  02/01/2024   
3      1372           Heladera  $114.35       8.0  02/01/2024   
4      1546           Secadora  $106.21       4.0  02/01/2024   
5      1697    Horno eléctrico   $35.35       9.0  02/01/2024   
6      1710   Plancha de vapor   $65.43       2.0  02/01/2024   
7      2959          Proyector   $88.17       9.0  02/01/2024   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos  
4  Electrodomésticos  
5  Electrodomésticos  
6  Electrodomésticos  
7        Electrónica  

id_venta         int64
producto        object
precio          object
cantidad       float64
fecha_venta     object
categoria       object
dtype: object


 Normalización de texto aplicad

## e) Optimización de los datos y memoria.

*   Una vez finalizada la normalización de los datos, la idea es primero transformar en "category" los datos que cumplan con las caracteristicas desedas ya analizado en el punto D). La columna "pruducto" de los DataFrames de Marketing y Ventas, si bien por lo que se pudo validar analizando los datos corresponde a agrupaciones de sub-productos, al parecer por su "especificación" no pareciera a priori categorizarla en category salvo que en análisis futuros sea evidente que esto podría optimizar nuevos reportes u operaciones.
*   Por ultimo, segun lo planteado tambien en el punto D), se reducirán los bytes los formatos especialmente de los datos numéricos a los optimos en función del espacio necesario por su rango de uso esperado.

### Código y verificación.

In [67]:
print("df_Clientes_v2 ANTES: \n")
df_Clientes_v2.info(memory_usage='deep')
df_Clientes_v2['ciudad'] = df_Clientes_v2['ciudad'].astype('category')
df_Clientes_v2['id_cliente'] = df_Clientes_v2['id_cliente'].astype('int16')
df_Clientes_v2['edad'] = df_Clientes_v2['edad'].astype('int8')
print("\n\ndf_Clientes_v2 DESPUES: \n")
df_Clientes_v2.info(memory_usage='deep')

df_Clientes_v2 ANTES: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 567 entries, 0 to 566
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id_cliente  567 non-null    int64  
 1   nombre      567 non-null    object 
 2   edad        567 non-null    int64  
 3   ciudad      567 non-null    object 
 4   ingresos    567 non-null    float64
dtypes: float64(1), int64(2), object(2)
memory usage: 80.8 KB


df_Clientes_v2 DESPUES: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 567 entries, 0 to 566
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   id_cliente  567 non-null    int16   
 1   nombre      567 non-null    object  
 2   edad        567 non-null    int8    
 3   ciudad      567 non-null    category
 4   ingresos    567 non-null    float64 
dtypes: category(1), float64(1), int16(1), int8(1), object(1)
memory usage: 42.7 KB


In [68]:
print("df_Marketing_v2 ANTES: \n")
df_Marketing_v2.info(memory_usage='deep')
df_Marketing_v2['canal'] = df_Marketing_v2['canal'].astype('category')
#df_Marketing_v2['producto'] = df_Marketing_v2['producto'].astype('category')
df_Marketing_v2['id_campanha'] = df_Marketing_v2['id_campanha'].astype('int16')
print("\n\df_Marketing_v2 DESPUES: \n")
df_Marketing_v2.info(memory_usage='deep')

df_Marketing_v2 ANTES: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   id_campanha   90 non-null     int64         
 1   producto      90 non-null     object        
 2   canal         90 non-null     object        
 3   costo         90 non-null     float64       
 4   fecha_inicio  90 non-null     datetime64[ns]
 5   fecha_fin     90 non-null     datetime64[ns]
dtypes: datetime64[ns](2), float64(1), int64(1), object(2)
memory usage: 13.0 KB

\df_Marketing_v2 DESPUES: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   id_campanha   90 non-null     int16         
 1   producto      90 non-null     object        
 2   canal         90 non-null     category      
 3   costo 

<>:6: SyntaxWarning: invalid escape sequence '\d'
<>:6: SyntaxWarning: invalid escape sequence '\d'
/tmp/ipython-input-3681162089.py:6: SyntaxWarning: invalid escape sequence '\d'
  print("\n\df_Marketing_v2 DESPUES: \n")


In [69]:
print("df_Ventas_v2 ANTES: \n")
df_Ventas_v2.info(memory_usage='deep')
#df_Ventas_v2['producto'] = df_Ventas_v2['producto'].astype('category')
df_Ventas_v2['categoria'] = df_Ventas_v2['categoria'].astype('category')
df_Ventas_v2['id_venta'] = df_Ventas_v2['id_venta'].astype('int16')
print("\n\df_Ventas_v2 DESPUES: \n")
df_Ventas_v2.info(memory_usage='deep')

df_Ventas_v2 ANTES: 

<class 'pandas.core.frame.DataFrame'>
Index: 2998 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   id_venta     2998 non-null   int64         
 1   producto     2998 non-null   object        
 2   precio       2998 non-null   float64       
 3   cantidad     2998 non-null   float64       
 4   fecha_venta  2998 non-null   datetime64[ns]
 5   categoria    2998 non-null   object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(2)
memory usage: 477.8 KB

\df_Ventas_v2 DESPUES: 

<class 'pandas.core.frame.DataFrame'>
Index: 2998 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   id_venta     2998 non-null   int16         
 1   producto     2998 non-null   object        
 2   precio       2998 non-null   float64       
 3   cantidad     2998 non-null

<>:6: SyntaxWarning: invalid escape sequence '\d'
<>:6: SyntaxWarning: invalid escape sequence '\d'
/tmp/ipython-input-2133339048.py:6: SyntaxWarning: invalid escape sequence '\d'
  print("\n\df_Ventas_v2 DESPUES: \n")


## f) Reporte final comparativo DataFrames originales Vs. DataFrames procesados.

In [70]:
reporte_calidad_global(DFs = [df_Clientes, df_Marketing, df_Ventas], nombres = ['df_Clientes', 'df_Marketing', 'df_Ventas'])
reporte_calidad_global(DFs = [df_Clientes_v2, df_Marketing_v2, df_Ventas_v2], nombres = ['df_Clientes_v2', 'df_Marketing_v2', 'df_Ventas_v2'])


DataFrame  Filas  Columnas  Nulos totales  Duplicados   Memoria
0   df_Clientes    567         5              0           0  0.082819
1  df_Marketing     90         6              0           0  0.022093
2     df_Ventas   3035         6              4          70  0.848279

DataFrame  Filas  Columnas  Nulos totales  Duplicados   Memoria
0   df_Clientes_v2    567         5              0           0  0.041654
1  df_Marketing_v2     90         6              0           0  0.007961
2     df_Ventas_v2   2998         6              0           0  0.276311

# I) Enriquesimiento de datos de los DataFrames limpios (v2) y creación de nuevos DataFrames auxiliares.
> El objetivo es agregar las columnas y/o DataFrames auxiliares antes de integrar la información con el objeto de facilitar "joins", dar mayor trazabilidad y auditoría al proceso, estandarizar agrupaciones, categorizaciones, etc.

> Se procede en base a lo analizado en D), con los DataFrames en su versión posterior a la limpieza y a la transformación en funcion de su calidad.

> Los DataFrames surgidos luego de este proceso se encontraran con la denominacion v3 luego del nombre del archivo.|

##Código y verificación.

In [71]:
# Armado de df_Clientes_v3
# Realizo una copia del DataFrame limpio y con procesos de claidad (V2)
df_Clientes_v3 = df_Clientes_v2.copy()
# Agrego columnas nuevas:

# Nueva columna rango_edad (Se procesa con edades/categorias con una estructura de tipo FIJO -categorias inamobibles-
# ya es un dato donde los rangos y sus significados se mantienen estables en el tiempo)
df_Clientes_v3['rango_edad'] = pd.cut(df_Clientes_v3['edad'], bins=[0,18,35,60,100], labels=['MENOR','JOVEN','ADULTO','MAYOR'])
df_Clientes_v3['rango_edad'] = df_Clientes_v3['rango_edad'].astype('category')

# Nueva columna rango_ingresos (Se procesa con rangos con una estructura DINAMICA -rangos representativos de un %, quantil, etc-
# ya es un dato donde los valores y sus significados pueden variar dependiendo del periodo estudiado)
# Se utiliza la función genérica para categorizar por cuantiles.
# Función utilizada: "categorizar_por_cuantiles".
df_Clientes_v3 = categorizar_por_cuantiles(
    DF=df_Clientes_v3,
    columna='ingresos',
    nombre_columna_nueva='rango_ingresos',
    nombre="df_Clientes_v3",
    num_cuantiles=4,
    etiquetas=['BAJO', 'MEDIO_BAJO', 'MEDIO_ALTO', 'ALTO']
)

# Verifico la Salida
print(f"\n df_Clientes_v3 DESPUES de la AGREGACION: \n{df_Clientes_v3.head(8)}")


Columna 'ingresos' del dataframe 'df_Clientes_v3' categorizada por cuantiles en 'rango_ingresos'.

 df_Clientes_v3 DESPUES de la AGREGACION: 
   id_cliente               nombre  edad                 ciudad  ingresos  \
0           1      Aloysia Screase    44          Mar Del Plata  42294.68   
1           2  Kristina Scaplehorn    25                Posadas  24735.04   
2           3       Filip Castagne    50            Resistencia  35744.85   
3           4          Liuka Luard    39           Bahia Blanca  27647.96   
4           5        Dore Cockshtt    28                Rosario  28245.65   
5           6        Patrick Earle    34  San Miguel De Tucuman  62763.31   
6           7           Etan Deeth    35            Resistencia  37489.71   
7           8       Booth Bielfelt    40                Cordoba  35255.94   

  rango_edad rango_ingresos  
0     ADULTO     MEDIO_ALTO  
1      JOVEN           BAJO  
2     ADULTO     MEDIO_ALTO  
3     ADULTO     MEDIO_BAJO  
4      JOVEN 

In [72]:
# Armado de df_Marketing_v3
# Realizo una copia del DataFrame limpio y con procesos de claidad (V2)
df_Marketing_v3 = df_Marketing_v2.copy()
# Agrego columnas nuevas:

# Columnas para Agrupar fechas
df_Marketing_v3['yyyy_mm_inicio'] = df_Marketing_v3['fecha_inicio'].dt.to_period('M')
df_Marketing_v3['yyyy_mm_fin'] = df_Marketing_v3['fecha_fin'].dt.to_period('M')

# Nueva columna duracion_campanha_dias (int16)
# Calculamos la diferencia en días y la convertimos a int16
df_Marketing_v3['duracion_campanha_dias'] = (df_Marketing_v3['fecha_fin'] - df_Marketing_v3['fecha_inicio']).dt.days.astype('int16')

# Control para deltas cero o negativos (basado en duracion_campanha_dias)
if (df_Marketing_v3['duracion_campanha_dias'] <= 0).any():
    print("\n¡Advertencia! Se encontraron campañas con duración cero o negativa.")
    display(df_Marketing_v3[df_Marketing_v3['duracion_campanha_dias'] <= 0])

# Nueva columna rango_duracion_campanha (Categorical based on days)
df_Marketing_v3['rango_duracion_campanha'] = pd.cut(df_Marketing_v3['duracion_campanha_dias'], bins=[0,30,60,90,360], labels=['MENSUAL','BIMESTRAL','TRIMESTRAL','LARGA_DURACION'], include_lowest=True, right=False)
df_Marketing_v3['rango_duracion_campanha'] = df_Marketing_v3['rango_duracion_campanha'].astype('category')

# Nueva columna costo_mensualizado
# Calculate the number of months in the campaign duration (handle potential division by zero or very short durations)
# Adding a small epsilon to avoid division by zero for durations less than a day
# Calcular el número de meses en la duración de la campaña (manejar la posible división por cero o duraciones muy cortas)
# Agregar un pequeño epsilon para evitar la división por cero para duraciones menores a un día
df_Marketing_v3['duracion_meses'] = (df_Marketing_v3['duracion_campanha_dias'] / 30.44).replace(0, np.nan) # Approximate number of days in a month # Número aproximado de días en un mes
df_Marketing_v3['costo_mensualizado'] = (df_Marketing_v3['costo'] / df_Marketing_v3['duracion_meses']).round(2) # Round to 2 decimal places # Redondear a 2 decimales
df_Marketing_v3 = df_Marketing_v3.drop(columns=['duracion_meses']) # Drop the intermediate column # Eliminar la columna intermedia

# Nueva columna rango_costo_mensualizado
df_Marketing_v3 = categorizar_por_cuantiles(
    DF=df_Marketing_v3,
    columna='costo_mensualizado',
    nombre_columna_nueva='rango_costo_mensualizado',
    nombre="df_Marketing_v3",
    num_cuantiles=4,
    etiquetas=['BAJO', 'MEDIO_BAJO', 'MEDIO_ALTO', 'ALTO']
)

# Por el momento descarto agregar una nueva columna 'categoria' respecto del
# 'producto' ya que parece que el uso que se le dará en el futuro esta más
# ligado a la desagregacion de la inforamacion producto que a integrarla en sus
# 'categoria'

# Verifico la Salida
print(f"\n df_Marketing_v3 DESPUES de la AGREGACION: \n{df_Marketing_v3.head(8)}")


Columna 'costo_mensualizado' del dataframe 'df_Marketing_v3' categorizada por cuantiles en 'rango_costo_mensualizado'.

 df_Marketing_v3 DESPUES de la AGREGACION: 
   id_campanha          producto  canal  costo fecha_inicio  fecha_fin  \
0           74   ADORNO DE PARED     TV   4.81   2024-03-20 2024-05-03   
1           12            TABLET   RRSS   3.40   2024-03-26 2024-05-13   
2           32   LAMPARA DE MESA  EMAIL   5.54   2024-03-28 2024-04-20   
3           21        SMARTPHONE   RRSS   6.37   2024-03-29 2024-05-16   
4           58          ALFOMBRA  EMAIL   4.25   2024-03-31 2024-05-05   
5           85        SMARTWATCH     TV   5.07   2024-04-01 2024-05-05   
6           36  PLANCHA DE VAPOR  EMAIL   5.41   2024-04-02 2024-06-01   
7           57          BATIDORA  EMAIL   4.48   2024-04-10 2024-06-08   

  yyyy_mm_inicio yyyy_mm_fin  duracion_campanha_dias rango_duracion_campanha  \
0        2024-03     2024-05                      44               BIMESTRAL   
1       

In [73]:
# Armado de df_Ventas_v3
# Realizo una copia del DataFrame limpio y con procesos de claidad (V2)
df_Ventas_v3 = df_Ventas_v2.copy()
# Agrego columnas nuevas:

# Columna para Agrupar fechas
df_Ventas_v3['yyyy_mm_venta'] = df_Ventas_v3['fecha_venta'].dt.to_period('M')

# Nueva columna importe_bruto_venta (precio * cantidad) float64
df_Ventas_v3['importe_bruto_venta'] = (df_Ventas_v3['precio'] * df_Ventas_v3['cantidad']).astype('float64')

# Nueva columna rango_importe_bruto_venta
df_Ventas_v3 = categorizar_por_cuantiles(
    DF=df_Ventas_v3,
    columna='importe_bruto_venta',
    nombre_columna_nueva='rango_importe_bruto_venta',
    nombre="df_Ventas_v3",
    num_cuantiles=4,
    etiquetas=['BAJO', 'MEDIO_BAJO', 'MEDIO_ALTO', 'ALTO']
)

# Verifico la Salida
print(f"\n df_Ventas_v3 DESPUES de la AGREGACION: \n{df_Ventas_v3.head(8)}")


Columna 'importe_bruto_venta' del dataframe 'df_Ventas_v3' categorizada por cuantiles en 'rango_importe_bruto_venta'.

 df_Ventas_v3 DESPUES de la AGREGACION: 
   id_venta           producto  precio  cantidad fecha_venta  \
0       792  CUADRO DECORATIVO   69.94       5.0  2024-01-02   
1       811    LAMPARA DE MESA  105.10       5.0  2024-01-02   
2      1156           SECADORA   97.96       3.0  2024-01-02   
3      1372           HELADERA  114.35       8.0  2024-01-02   
4      1546           SECADORA  106.21       4.0  2024-01-02   
5      1697    HORNO ELECTRICO   35.35       9.0  2024-01-02   
6      1710   PLANCHA DE VAPOR   65.43       2.0  2024-01-02   
7      2959          PROYECTOR   88.17       9.0  2024-01-02   

           categoria yyyy_mm_venta  importe_bruto_venta  \
0         DECORACION       2024-01               349.70   
1         DECORACION       2024-01               525.50   
2  ELECTRODOMESTICOS       2024-01               293.88   
3  ELECTRODOMESTICOS      

## Anexo Especial

### a) Tabla productos alto rendimiento.

In [74]:
# Creación del DataFrame auxiliar de productos de alto rendimiento
df_aux_prod_alto_rendimiento_v3 = df_Ventas_v3[df_Ventas_v3['rango_importe_bruto_venta'] == 'ALTO'].copy()

# Ordenar por importe_bruto_venta de mayor a menor
df_aux_prod_alto_rendimiento = df_aux_prod_alto_rendimiento_v3.sort_values(by='importe_bruto_venta', ascending=False)

# Verifico la Salida
print(f"\n DataFrame auxiliar df_aux_prod_alto_rendimiento_v3 creado TOP 20:\n\n{df_aux_prod_alto_rendimiento_v3.head(20)}")


 DataFrame auxiliar df_aux_prod_alto_rendimiento_v3 creado TOP 20:

    id_venta                producto  precio  cantidad fecha_venta  \
3       1372                HELADERA  114.35       8.0  2024-01-02   
7       2959               PROYECTOR   88.17       9.0  2024-01-02   
8        318       RINCON DE PLANTAS   79.86      11.0  2024-01-03   
19       882       CUADRO DECORATIVO  101.92       9.0  2024-01-04   
20       969              CANDELABRO   90.70      12.0  2024-01-04   
22      1971      FREIDORA ELECTRICA  122.25      11.0  2024-01-04   
25       606                CORTINAS  106.66      11.0  2024-01-05   
27      1362                HELADERA  109.49      11.0  2024-01-05   
28      2155             AURICULARES   84.80      10.0  2024-01-05   
32      2180  CONSOLA DE VIDEOJUEGOS  102.57       9.0  2024-01-06   
36        17   ELEMENTOS DE CERAMICA  108.48      11.0  2024-01-07   
43      1496                SECADORA   95.33      11.0  2024-01-07   
46      2174         

### a) Analisis de ingresos brutos por ventas segun el producto y sus caracteristicas.

In [75]:
# Creación del DataFrame auxiliar solicitado.
df_resumen_ingreso_producto_v3 = (
    df_Ventas_v3
    .groupby(
        by='producto',
        dropna=False,
        as_index=False,
        observed=False
    )
    .agg(
        ingreso_total=('importe_bruto_venta', 'sum'),
        unidades=('cantidad', 'sum'),
        precio_promedio=('precio', 'mean'),
        registros=('importe_bruto_venta', 'size')       #
    )
)

# Redondear precio_promedio a 2 decimales
df_resumen_ingreso_producto_v3['precio_promedio'] = df_resumen_ingreso_producto_v3['precio_promedio'].round(2)

# Nueva columna rango_ingreso_total usando categorizar_por_cuantiles
df_resumen_ingreso_producto_v3 = categorizar_por_cuantiles(
    DF=df_resumen_ingreso_producto_v3,
    columna='ingreso_total',
    nombre_columna_nueva='rango_ingreso_total',
    nombre="df_resumen_ingreso_producto_v3",
    num_cuantiles=4,
    etiquetas=['BAJO', 'MEDIO_BAJO', 'MEDIO_ALTO', 'ALTO']
)

# Verifico la Salida
print(f"\n DataFrame auxiliar df_resumen_ingreso_producto_v3 :\n\n")
display(df_resumen_ingreso_producto_v3)


Columna 'ingreso_total' del dataframe 'df_resumen_ingreso_producto_v3' categorizada por cuantiles en 'rango_ingreso_total'.

 DataFrame auxiliar df_resumen_ingreso_producto_v3 :




producto  ingreso_total  unidades  precio_promedio  \
0          ADORNO DE PARED       48093.49     633.0            76.10   
1                 ALFOMBRA       44773.06     615.0            74.10   
2               ASPIRADORA       50085.86     651.0            77.45   
3              AURICULARES       74175.58     958.0            76.30   
4                 BATIDORA       50979.20     672.0            77.54   
5                 CAFETERA       59607.31     765.0            79.05   
6           CAMARA DIGITAL       45217.96     638.0            72.01   
7               CANDELABRO       11128.80     160.0            74.60   
8   CONSOLA DE VIDEOJUEGOS       46174.41     623.0            76.35   
9                 CORTINAS       44865.03     610.0            73.65   
10       CUADRO DECORATIVO       54297.60     726.0            74.58   
11   ELEMENTOS DE CERAMICA       45411.09     636.0            71.61   
12       ESPEJO DECORATIVO       46783.31     635.0            75.66   
13      FREIDORA ELECTRICA       50155.15     630.0            78.25   
14                HELADERA       25736.80     330.0            74.13   
15         HORNO ELECTRICO       49913.90     633.0            75.69   
16       JARRON DECORATIVO       51130.88     672.0            74.53   
17         LAMPARA DE MESA       82276.38    1112.0            72.72   
18                  LAPTOP       46179.48     617.0            75.33   
19                LAVADORA       48946.44     671.0            74.95   
20              MICROONDAS       72562.89     912.0            79.18   
21     PARLANTES BLUETOOTH       47779.46     660.0            73.58   
22        PLANCHA DE VAPOR       45196.63     632.0            73.04   
23               PROYECTOR       48187.00     626.0            74.68   
24       RINCON DE PLANTAS       50456.45     691.0            74.82   
25                SECADORA       52115.45     696.0            73.34   
26              SMARTPHONE       54132.44     665.0            81.40   
27              SMARTWATCH       47836.86     620.0            75.54   
28                  TABLET       25856.57     374.0            71.38   
29               TELEVISOR       47038.04     632.0            73.82   

    registros rango_ingreso_total  
0         100          MEDIO_BAJO  
1         100                BAJO  
2         100          MEDIO_ALTO  
3         143                ALTO  
4         100          MEDIO_ALTO  
5         117                ALTO  
6         100                BAJO  
7          24                BAJO  
8          99          MEDIO_BAJO  
9         100                BAJO  
10        100                ALTO  
11         99                BAJO  
12        100          MEDIO_BAJO  
13        100          MEDIO_ALTO  
14         48                BAJO  
15        100          MEDIO_ALTO  
16        100                ALTO  
17        176                ALTO  
18        100          MEDIO_BAJO  
19        100          MEDIO_ALTO  
20        135                ALTO  
21         98          MEDIO_BAJO  
22        100                BAJO  
23         99          MEDIO_ALTO  
24        101          MEDIO_ALTO  
25        100                ALTO  
26        101                ALTO  
27        100          MEDIO_BAJO  
28         58                BAJO  
29        100          MEDIO_BAJO

# J) Integración de Datos los DataFrames.
> Se evaluara en la proxima entrega

# K) Conversion final de los DataFrame modificados a CSV.


In [76]:
from google.colab import drive
import os

# Montar Google Drive
drive.mount('/content/drive')
dir_salida = '/content/drive/MyDrive/datasets/datos_procesados_pre_entrega'

# Crear el directorio si no existe
if not os.path.exists(dir_salida):
    os.makedirs(dir_salida)
    print(f"Directorio '{dir_salida}' creado.")
else:
    print(f"Directorio '{dir_salida}' ya existe.")

# Guardar los DataFrames terminados en _v3
dataframes_a_guardar = {
    'df_Clientes_v3': df_Clientes_v3,
    'df_Marketing_v3': df_Marketing_v3,
    'df_Ventas_v3': df_Ventas_v3,
    'df_aux_prod_alto_rendimiento_v3': df_aux_prod_alto_rendimiento_v3,
    'df_resumen_ingreso_producto_v3': df_resumen_ingreso_producto_v3
}

for name, df in dataframes_a_guardar.items():
    file_path = os.path.join(dir_salida, f'{name}.csv')
    df.to_csv(file_path, index=False)
    print(f"DataFrame '{name}' guardado en '{file_path}'")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Directorio '/content/drive/MyDrive/datasets/datos_procesados_pre_entrega' ya existe.
DataFrame 'df_Clientes_v3' guardado en '/content/drive/MyDrive/datasets/datos_procesados_pre_entrega/df_Clientes_v3.csv'
DataFrame 'df_Marketing_v3' guardado en '/content/drive/MyDrive/datasets/datos_procesados_pre_entrega/df_Marketing_v3.csv'
DataFrame 'df_Ventas_v3' guardado en '/content/drive/MyDrive/datasets/datos_procesados_pre_entrega/df_Ventas_v3.csv'
DataFrame 'df_aux_prod_alto_rendimiento_v3' guardado en '/content/drive/MyDrive/datasets/datos_procesados_pre_entrega/df_aux_prod_alto_rendimiento_v3.csv'
DataFrame 'df_resumen_ingreso_producto_v3' guardado en '/content/drive/MyDrive/datasets/datos_procesados_pre_entrega/df_resumen_ingreso_producto_v3.csv'
